interactive/printing: Add support for MutableNDimArray for printing

peey opened the pull request:
NDimArrays of the mutable type had to be explicitly added to `_can_print_latex` function and `printable_types` array so that they would display properly in Latex (in an IPython / Jupyter environment)

Commit 2 does exactly this. Commit 3 improves on this by merging the shared logic of `_can_print_latex` and `printable_types` to remove duplication, and Commit 1 is removal of dead code (general housekeeping).

Fixes #12386

peey committed:
tensor/array/ndim_array: remove dead code from __str__

The __str__ function had some code after the return statement which
I removed

peey committed:
interactive/printing: Make NDimArray printable in latex

There were problems with latex printing for MutableDenseNDimArray, and
MutableSparseNDimArray. This commit fixes that issue by adding
NDimArray to printable_types and _can_print_latex

peey committed:
interactive/printing: remove duplication in printable_types

printable_types and the _can_print_latex function contained duplicate
code which I refactored to remove duplication

siefkenj commented on sympy/interactive/printing.py:
```diff
@@ -96,28 +96,25 @@ def _matplotlib_wrapper(o):
-            return True
-        return False
+        printable_types = _get_latex_printable_types()
+        return isinstance(o, tuple(printable_types))
```
Aren't you missing a lot of logic from before?

```
x = (1,2,3,MyUnprintableType)
_can_print_latex(x)
```
would return `False` before, but now it returns `True`.  Unless I am reading incorrectly.

peey replied:
Ah. I didn't realize that for the first input, the operator could be a tuple. Your example is correct.

peey committed:
interactive/printing: undo previous commit about printable_types

The logic introduced in previous commit (8b92f57d5) was erroneous.
Revert to the state of the file incommit before it (ed211a3)

siefkenj commented:
Does `NDimArray` only contain printable types, or is it possible it could contain unprintable types?

siefkenj commented on sympy/interactive/printing.py:
```diff
@@ -117,6 +121,8 @@ def _can_print_latex(o):
             return True
```
If `NDimArray` is like `Matrix` in that its elements are assumed printable, you should just add it to this list instead of creating a new `elif` clause.

peey replied:
Yeah, it makes sense to group the similar ones in that clause.

peey commented:
> Does NDimArray only contain printable types, or is it possible it could contain unprintable types?

I think it would be a safe assumption that it only contains printable types. The reason behind this is that the Immutable subclasses of NDimArray already inherit from Basic and were considered printable even before this commit. 

This commit just makes Mutable subclasses printable, for which printing functions were already defined in printing/latex.py

peey committed:
interactive/printing: refactor elif clause in _can_print_latex

Move NDimArray from a separate elif clause to the clause dealing
with all other types for which assumption that all children are
printable is made

siefkenj commented:
This looks good!  Can you add a test for it?

As a separate issue, it seems that there is nothing in the tests that tests printing of a tuple containing an unprintable type.  Perhaps adding such a test would be a nice PR.

peey commented:
I'll work on that. This code should have some tests, that would have been a good first pointer towards an error when I made that refactor commit which I had to revert. 

I want to work on adding unit tests for the function `_can_print_latex` but it is an inner function. This would require modifications in the code structure. I guess I would bring it out. I'm not sure if the other inner functions in the file are testable so I'll leave them be.

siefkenj commented:
I agree.  At a minimum, the `def _can_print_latex` would have to be moved out of the body of the function.  If I had a better understanding of the printing module, I suspect there is some reuse that could be going on.

In any case, this is probably better as a separate PR

peey commented:
Alright then. I'll work on that as a separate PR and open a separate issue for it. I'll request for this PR to be merged since I have your approval on this.

peey commented:
The commit is failing tests with seemingly no relation to the code that was changed. [Job #21416.32](https://travis-ci.org/sympy/sympy/jobs/213839851) fails for wrong example in dpll2, and [Job #21416.21](https://travis-ci.org/sympy/sympy/jobs/213839840), [Job #21416.21](https://travis-ci.org/sympy/sympy/jobs/213839841) for `def test_DenseMatrix():` in `test_theanocode.py`

I'm not sure what to do
